In [1]:
import numpy as np
import pandas as pd
import os
import matplotlib.pyplot as plt
% matplotlib inline

In [2]:
os.listdir(os.getcwd())

['.ipynb_checkpoints',
 '100-pandas-puzzles.ipynb',
 '11DSHW2_linear_regression_with_cross_validation.ipynb',
 '11DS_HW1.ipynb',
 '11DS_HW3.ipynb',
 'ch08.pdf',
 'ch09.pdf',
 'DAL_HW3',
 'data structure and algorithms in python ch1-3.ipynb',
 'data structure and algorithms in python ch4.ipynb',
 'Lecture1.ipynb',
 'LeetCode.ipynb',
 'machine learning in python.ipynb',
 'ML_hw1.pdf',
 'ML_hw1.tex',
 'ML_hw2.aux',
 'ML_hw2.log',
 'ML_hw2.pdf',
 'ML_hw2.synctex.gz',
 'ML_hw2.tex',
 'model_selection.cross_val_score.ipynb',
 'nda_nov_ds.pdf',
 'Np_exercises.ipynb',
 'Roc+curve.ipynb',
 'roc_curve.png',
 'train.csv',
 'training_contract.pdf',
 'Untitled.ipynb',
 'Untitled1.ipynb',
 'Week1-Notebook.ipynb',
 'Week2Notebook.ipynb',
 'Week3Notebook.ipynb']

In [5]:
data = pd.read_excel('DAL_HW3\NEISS2014.xlsx')

In [6]:
data.head()

CPSC Case #  trmt_date  psu  weight stratum  age     sex         race  \
0    141200216 2014-10-23   63  99.704       M   21  Female  None listed   
1    140117851 2014-01-04   63  81.576       M   62  Female        White   
2    150144993 2014-08-17   63  99.704       M   21  Female  None listed   
3    150230176 2014-07-27   63  99.704       M   30  Female  None listed   
4    141220717 2014-09-08   63  99.704       M   16    Male  None listed   

  race_other  diag diag_other  body_part  disposition  location  fmv  prod1  \
0        NaN    62        NaN         75            1         1    0    679   
1        NaN    57        NaN         79            1         1    0   1807   
2        NaN    57        NaN         83            1         0    0   1333   
3        NaN    64        NaN         79            1         1    0   4076   
4        NaN    57        NaN         82            1         0    0   1893   

    prod2                                          narrative  
0  1807.0  21YOF LAYING ON THE COUCH AND ROLLED OFF TO TH...  
1   115.0  62YOF TRIPPED OVER THE CORD FOR THE VACUUM CLE...  
2     NaN    21YOF FELL OFF A SKATEBOARD AND FRACTURED FOOT   
3     NaN  30YOF TRYING TO GET OUT OF BED AND FELT A POP ...  
4     NaN   16YOM HAD HAND SLAMMED IN A DOOR FRACTURED HAND

# 数据应用学院Data Scientist Program Hw3

#### The 'NEISS' dataset contains some injury information. Dataset along with coding manual can be downloaded from Piazza.
#### Feel free to explore and share with us any insights you may get from this dataset.

#### 1. What are the top three body parts most frequently represented in this dataset? What are the top three body parts that are least frequently represented?

In [7]:
body_part_freq = data.body_part.value_counts()

In [8]:
print body_part_freq[0:3]
print body_part_freq[-3:]

75    9891
76    5786
92    5783
Name: body_part, dtype: int64
87    390
38    286
84      4
Name: body_part, dtype: int64


#### Checking the manual, the three most frequently represented body parts are:
75: Head, 76: Face, 92: Finger
#### and the three least frequently represented body parts are:
84: 25-50% of body, 38: Pubic region, 87: Not stated

#### 2. How many injuries in this dataset involve a skateboard?
#### Of those injuries, what percentage were male and what percentage were female? What was the average age of someone injured in an incident involving a skateboard?

In [9]:
skate_data = data[(data.prod1 == 1333) | (data.prod2 == 1333)]

In [10]:
len(skate_data)

495

#### there are 495 injuries that involve skateboards

In [11]:
from __future__ import division

In [12]:
np.sum(skate_data.sex == 'Female') / 495

0.17979797979797979

In [13]:
np.sum(skate_data.sex == 'Male') / 495

0.82020202020202015

In [14]:
skate_data.age.mean()

17.89090909090909

#### 18% of these incidents happened on female and 82% of them on male, the average age is 17.9

#### 3. What diagnosis had the highest hospitalization rate?
#### What diagnosis most often concluded with the individual leaving without being seen?
#### Briefly discuss your findings and any caveats you'd mention when discussing this data

In [15]:
grouped_diag = data.groupby(['diag'])

In [16]:
df = grouped_diag.disposition.value_counts()

In [17]:
diag_set = data.diag.unique()
diag_set

array([62, 57, 64, 72, 59, 53, 48, 71, 67, 55, 41, 74, 51, 56, 54, 63, 65,
       52, 66, 68, 58, 60, 49, 42, 50, 61, 73, 69, 47, 46], dtype=int64)

In [18]:
df1 = pd.DataFrame(np.zeros((len(diag_set), 2)), index = diag_set, columns=['hospitalization_rate', 'unseen_rate'])
df1

hospitalization_rate  unseen_rate
62                   0.0          0.0
57                   0.0          0.0
64                   0.0          0.0
72                   0.0          0.0
59                   0.0          0.0
53                   0.0          0.0
48                   0.0          0.0
71                   0.0          0.0
67                   0.0          0.0
55                   0.0          0.0
41                   0.0          0.0
74                   0.0          0.0
51                   0.0          0.0
56                   0.0          0.0
54                   0.0          0.0
63                   0.0          0.0
65                   0.0          0.0
52                   0.0          0.0
66                   0.0          0.0
68                   0.0          0.0
58                   0.0          0.0
60                   0.0          0.0
49                   0.0          0.0
42                   0.0          0.0
50                   0.0          0.0
61                   0.0          0.0
73                   0.0          0.0
69                   0.0          0.0
47                   0.0          0.0
46                   0.0          0.0

In [19]:
for i in diag_set:
    total = df[i].sum()
    try:
        hospitalization = df[i][4]
    except KeyError:
        hospitalization = 0
    try:
        unseen = df[i][6]
    except KeyError:
        unseen = 0
    df1.loc[i, 'hospitalization_rate'] = hospitalization / total
    df1.loc[i, 'unseen_rate'] = unseen / total

In [20]:
df1.sort_values('hospitalization_rate', ascending = False)

hospitalization_rate  unseen_rate
69              0.425926     0.000000
57              0.186030     0.000616
50              0.160000     0.016000
68              0.121547     0.033149
62              0.118734     0.011873
41              0.115942     0.002070
47              0.105263     0.000000
67              0.093750     0.000000
61              0.085859     0.000000
58              0.078652     0.022472
65              0.076471     0.017647
42              0.075758     0.030303
71              0.075493     0.031650
48              0.073980     0.012755
51              0.052734     0.019531
52              0.046823     0.004682
55              0.043522     0.003036
72              0.035608     0.005935
63              0.023585     0.011006
53              0.020383     0.004884
66              0.015625     0.000000
59              0.015032     0.009994
60              0.007968     0.015936
56              0.007874     0.007874
74              0.004831     0.004831
64              0.004164     0.002034
54              0.000000     0.000000
49              0.000000     0.023256
73              0.000000     0.028571
46              0.000000     0.000000

In [21]:
df1.sort_values('unseen_rate', ascending = False)

hospitalization_rate  unseen_rate
68              0.121547     0.033149
71              0.075493     0.031650
42              0.075758     0.030303
73              0.000000     0.028571
49              0.000000     0.023256
58              0.078652     0.022472
51              0.052734     0.019531
65              0.076471     0.017647
50              0.160000     0.016000
60              0.007968     0.015936
48              0.073980     0.012755
62              0.118734     0.011873
63              0.023585     0.011006
59              0.015032     0.009994
56              0.007874     0.007874
72              0.035608     0.005935
53              0.020383     0.004884
74              0.004831     0.004831
52              0.046823     0.004682
55              0.043522     0.003036
41              0.115942     0.002070
64              0.004164     0.002034
57              0.186030     0.000616
47              0.105263     0.000000
69              0.425926     0.000000
67              0.093750     0.000000
61              0.085859     0.000000
54              0.000000     0.000000
66              0.015625     0.000000
46              0.000000     0.000000

In [35]:
a = pd.Series([1, 2, 3, 4])
b = pd.Series([3, 4, 5])
a / b

0    0.333333
1    0.500000
2    0.600000
3         NaN
dtype: float64

In [40]:
# alternative method for computing the rates

data[data.disposition == 4].groupby('diag').size().divide(data.groupby('diag').size()).sort_values(ascending = False)

diag
69    0.425926
57    0.186030
50    0.160000
68    0.121547
62    0.118734
41    0.115942
47    0.105263
67    0.093750
61    0.085859
58    0.078652
65    0.076471
42    0.075758
71    0.075493
48    0.073980
51    0.052734
52    0.046823
55    0.043522
72    0.035608
63    0.023585
53    0.020383
66    0.015625
59    0.015032
60    0.007968
56    0.007874
74    0.004831
64    0.004164
46         NaN
49         NaN
54         NaN
73         NaN
dtype: float64

#### Diag 69, which correspond to submersion, has highest hospitalization rate.  Diag 68, which correspond to poisioning, has most frequent people left unseen. This result makes sense, for submersion(including drowning) is always emergent and requires immediate treatment while poisioning food is less likely to be emergent and people may left themselves when they feel a little more comfortable later.


#### 4. Visualize any existing relationship between age and reported injuries (open question)


In [ ]:
from sklearn import linear_model
clf = linear_model.LogisticRegression(multi_class = 'ovr')

In [ ]:
X = data['age']
Y = data['body_part']

In [ ]:
X = X.values.reshape((-1, 1))

In [ ]:
clf.fit(X, Y)

In [ ]:
from sklearn import metrics

In [ ]:
metrics.accuracy_score(Y, clf.predict(X))